src/levanter/mesh.py

```diff
@@ -8,29 +8,51 @@
 def local_device_grid_positions(mesh, process_index: Optional[int] = None) -> tuple[np.ndarray, np.ndarray]:
     """Returns a tuple of nd arrays, one for each axis, indicating the position of each device on the grid.
     Analogous to what np.where would return."""
-    pi = process_index or jax.process_index()
-    # our device indices are [process_index * num_devices_per_node, (process_index + 1) * num_devices_per_node)
-    # we could be clever here and do math to figure out where we are in the grid, but it's simpler and less
-    # fragile to just search the grid for our devices
-    my_device_pos = np.vectorize(lambda dev: dev.process_index == pi)(mesh.devices)
+    if process_index is None:
+        process_index = jax.process_index()
+
+    my_device_pos = np.vectorize(lambda dev: dev.process_index == process_index)(mesh.devices)
     return my_device_pos.nonzero()
 
 
-def process_mesh_position(mesh, process_index: Optional[int] = None) -> tuple[int, ...]:
+def local_devices_mapping(mesh: Mesh, process_index: Optional[int] = None) -> dict[int, int]:
     """
-    If we envision each process as a subgrid of the mesh for its devices, this is the position of the process
-    in the coarsened process-level mesh
+    Handles the case when different devices in same process share the same data in TP.
+    Returns a mapping from local devices' DP/FSDP group index in global mesh to local indices
     """
-    upper_left_position = np.array([np.min(axis) for axis in local_device_grid_positions(mesh, process_index)])
-    local_mesh_size = mesh.local_mesh.devices.shape
-    pos = upper_left_position // local_mesh_size
-    return pos
+    local_device_pos = local_device_grid_positions(mesh, process_index)[:2]  # first 2 axes are DP axes.
+    result = {}
+    uid = 0
+    for local_device_index in range(len(local_device_pos[0])):
+        key = local_device_pos[0][local_device_index] * mesh.devices.shape[1] + local_device_pos[1][local_device_index]
+        if key not in result:
+            # when two devices maps to the same key (different TP index), they will get the same data
```

Comment:
wait, if the TP index is different they get the same data?

Reply:
Yes. If they get the same DP/FSDP id, they should be getting the same data.

Reply:
oh i see